### ISSUE TITLE
Pod stuck in ContainerCreating: Unit ...slice already exists

### ISSUE THREAD

**Issue by kolyshkin** [kubernetes/kubernetes#102676]
#### What happened:

Errors like this one 

> May 27 06:38:19.960408 ip-10-0-220-230 hyperkube[1448]: E0527 06:38:19.960361    1448 pod_workers.go:190] "Error syncing pod, skipping" err="failed to ensure that the pod: 5ac83c3f-0b16-4cf2-a3cb-f67c19cd0e16 cgroups exist and are correctly applied: failed to create container for [kubepods burstable pod5ac83c3f-0b16-4cf2-a3cb-f67c19cd0e16] : Unit kubepods-burstable-pod5ac83c3f_0b16_4cf2_a3cb_f67c19cd0e16.slice already exists." pod="openshift-machine-config-operator/machine-config-daemon-mm7gt" podUID=5ac83c3f-0b16-4cf2-a3cb-f67c19cd0e16

(when using cgroupDriver: systemd)

#### What you expected to happen:

No such errors

#### How to reproduce it (as minimally and precisely as possible):

I don't know for sure.

#### Anything else we need to know?:

This was introduced in k8s in #102147 and backported to 1.21 in #102196, so needs to be fixed in both master and `release-1.21`.

RH BZ: https://bugzilla.redhat.com/show_bug.cgi?id=1965545

The cause is a regression in runc/libcontainer: https://github.com/opencontainers/runc/issues/2996

The fix is in https://github.com/opencontainers/runc/pull/2997, which should make its way into runc 1.0.0 GA.

Currently there is DNM PR to bump runc to the version with the fix: https://github.com/kubernetes/kubernetes/pull/102508, but we have decided (https://github.com/kubernetes/kubernetes/pull/102250#issuecomment-855922711) to wait until the release.

#### Environment:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

**Comment by odinuge** [comment]
You should probably mention that this only affect people who use `cgroupDriver: systemd`.

**Comment by Jeansen** [comment]
I have this problem when creating a fresh new cluster. With version 1.21.1 all worked fine. with 1.21.2 it no longer passes preflight. No containers are created, at all.

**Comment by Jeansen** [comment]
I just confirmed in my setup. Installing kublet, kubeadm  (and kubectl) in version 1.21.1-00, all works as expected. But with the current version 1.21.2-00 I cannot successfully init a cluster with kubeadm (because not containers have been started).

**Comment by liggitt** [comment]
cc @mattysweeps for 1.21 impact to systemd cgroup drivers

**Comment by chrischdi** [comment]
We can also confirm to not have the issue in `v1.21.0` but issue occured since using `v1.21.2` :-)

**Comment by dims** [comment]
/milestone v1.22

+1 to backport to 1.21 as well to fix the bug in 1.21.x

**Comment by kolyshkin** [comment]
1.21 backport: https://github.com/kubernetes/kubernetes/pull/103392

**Comment by chrischdi** [comment]
Should this get reopened as the fix will get reverted in #103483 ?

**Comment by odinuge** [comment]
We can keep this open as long as this is still an issue in v1.21.
/reopen

**Comment by ehashman** [comment]
@kolyshkin what's the status of this issue? I don't currently see any linked PRs for the 1.22 milestone.

**Comment by kolyshkin** [comment]
> what's the status of this issue? I don't currently see any linked PRs for the 1.22 milestone.

The fix (for the bug that reverted runc 1.0.0 bump) is merged (https://github.com/opencontainers/runc/pull/3081) to main branch, the backport to 1.0 (https://github.com/opencontainers/runc/pull/3085) is being reviewed as I type and I think will be merged shortly. Once that happens, I'll create a WIP PR to update runc to the tip of runc release-1.0 branch. Once 1.0.1 is released (I hope this week, see https://github.com/opencontainers/runc/issues/3076), I'll finalize the PR and do the backport as well.

**Comment by odinuge** [comment]
I still think https://github.com/kubernetes/kubernetes/pull/102250 fixes this issue in a safer and simpler way. It still need review, but it should mitigate this issue, and fix issues with how we deal with systemd. This fix should be done either way imo. I am not a big fan of merging big dependency bumps after code freeze, together with cherry-picking them, when there are simpler and easier ways to do it.

This bug is quite nasty, and makes k8s unusable with systemd as cgroup driver, both for `master` and the `v1.21` releases... :/

The required code has not yet been merged into runc upstream, so I guess it can take a while to get a release cut. The PR will also be quite big, and we have now seen bad regressions in two runc bumps, so I think we will have to review it quite thoroughly. We are also quite behind code freeze, and with the risk involved, I think doing such a big dependency bump before a new release is quite dangerous. At least when CI is still not in a good shape. The current runc diff to master: https://github.com/opencontainers/runc/compare/v1.0.0-rc95...master 

Any thoughts about the way forward with this @dims @ehashman @liggitt?

**Comment by dims** [comment]
I agree with @odinuge, the churn+bugs in runc is very concerning, if #102250 works, then we should just take that and hold off on the huge dependency bump.

cc @derekwaynecarr @SergeyKanzhelev as well.

**Comment by kolyshkin** [comment]
> The current runc diff to master: https://github.com/opencontainers/runc/compare/v1.0.0-rc95...master

This is not the right diff to look at. We have already upgraded to 1.0.0 and it caused an issue (https://github.com/kubernetes/kubernetes/pull/103483) causing a revert. The issue has been
fixed and backported to release-1.0 branch; we're going to have a release this week.

So, the diff to look at is the difference between 1.0.0 and to-be-released 1.0.1: https://github.com/opencontainers/runc/compare/v1.0.0...release-1.0

The list of PRs for 1.0.1 can be seen at https://github.com/opencontainers/runc/issues/3076; from the kubernetes perspective, this is limited to the fix that caused a revert of runc 1.0.0 bump.

The runc 1.0.0 (or 1.0.1) bump would also fix "dbus: connection closed by user" issue caused by dbus restart (reported to https://bugzilla.redhat.com/show_bug.cgi?id=1941456), but I guess it's not super critical (the workaround is to restart kubernetes).

I am fine either way, but bumping runc to 1.0.1 and backporting may be beneficial and should not introduce any new regressions as we have already tested 1.0.0.

**Comment by odinuge** [comment]
> This is not the right diff to look at. We have already upgraded to 1.0.0 and it caused an issue (#103483) causing a revert. The issue has been
fixed and backported to release-1.0 branch; we're going to have a release this week.

Oh, we are probably talking about different PRs then. We still need https://github.com/opencontainers/runc/pull/3082, and it is targeted to `v1.1.0`, so I assume we need to wait for that. In case we freeze `kubepods.slice`, and users use runc `v1.0.0` (or other runtimes with the same logic), containers can end up being permanently frozen. As there is no need to freeze those control groups in the first place, imo. we shouldn't do it.

If we decide to go for that, and keep freezing the control groups, we should look into how it impact things, so we don't cause issues to users; especially before cherry-picking to `v1.21`. But yeah, it is doable. (But by that time we can probably get runc `v1.1.0` instead, if that is then released)

> The list of PRs for 1.0.1 can be seen at opencontainers/runc#3076; from the kubernetes perspective, this is limited to the fix that caused a revert of runc 1.0.0 bump.

Yeah. But there might be other changes between `rc95` and `v1.0.0` that impact other things, since it didn't get that many days worth of testing. CI is also lacking, especially on the systemd cgroup driver, but it is being worked on.

> The runc 1.0.0 (or 1.0.1) bump would also fix "dbus: connection closed by user" issue caused by dbus restart (reported to https://bugzilla.redhat.com/show_bug.cgi?id=1941456), but I guess it's not super critical (the workaround is to restart kubernetes).

Ohh. I have not seen that issue before.. Can't seem to find an issue now either; do you know if there is one?. What versions is that issue impacting (eg. do we need to fix it in even older releases)? Guess we need to handle/track that separately either way... I don't have time to look into it right now.

> I am fine either way, but bumping runc to 1.0.1 and backporting may be beneficial and should not introduce any new regressions as we have already tested 1.0.0.

We have to bump runc once in a while either way, so merging other PRs does not mean we pin runc permanently. Bumping runc in the normal merge window gives time to test it (as we have seen before) to make sure we don't break things. :smile:

**Comment by ehashman** [comment]
/remove-priority important-soon
/priority critical-urgent
/label release-blocker

This is a serious regression in the systemd driver that affects 1.21.x/1.22 and should be considered a release blocker.

**Comment by dims** [comment]
> So, the diff to look at is the difference between 1.0.0 and to-be-released 1.0.1: [opencontainers/runc@v1.0.0...release-1.0](https://github.com/opencontainers/runc/compare/v1.0.0...release-1.0)

Aha! 

> The list of PRs for 1.0.1 can be seen at [opencontainers/runc#3076](https://github.com/opencontainers/runc/issues/3076); from the kubernetes perspective, this is limited to the fix that caused a revert of runc 1.0.0 bump.

That's good news!

> I am fine either way, but bumping runc to 1.0.1 and backporting may be beneficial and should not introduce any new regressions as we have already tested 1.0.0.

Your info definitely makes me feel better about going to 1.0.1 @kolyshkin

**Comment by ehashman** [comment]
Given the discussion in the last few comments, and given that https://github.com/opencontainers/runc/pull/3082 isn't yet merged in upstream runc, my inclination would be to bump runc to 1.0.1 as soon as possible for the longest soak time, and handle the other regression fixes as possible backports. WDYT?

**Comment by kolyshkin** [comment]
> Oh, we are probably talking about different PRs then. We still need opencontainers/runc#3082, and it is targeted to v1.1.0, so I assume we need to wait for that. 

@odinuge I did not see it as that critical, thought of it more as some nice-to-have but not really required-right-now set of fixes.

Can you elaborate on what do you think is the exact problem that opencontainers/runc#3082 fixes for kubernetes? Temporarily freezing the sub-cgroups of the pod? If this is it, than it's not a new issue -- it is there since v1.0.0-rc91~48 (https://github.com/opencontainers/runc/pull/2391/commits/afe83489d4243800cb68b5c766831a4ceaff4578).

If it is needed, we can definitely backport it to 1.0 and speed up the review. Could you please add a comment to that PR saying you think it needs to be backported to 1.0?

Another candidate is https://github.com/opencontainers/runc/pull/3067 -- please let us know if you feel it's needed.

If this is not the best place to discuss it, let's move on to https://github.com/opencontainers/runc/issues/3076

**Comment by kolyshkin** [comment]
Forgot to add -- I think #102250 makes sense, too.

**Comment by sfxworks** [comment]
Wow I switched from runc to crun thinking it would fix this just to discover this. Very odd.

**Comment by sfxworks** [comment]
I found this thread from https://bugzilla.redhat.com/show_bug.cgi?id=1466636 
I get a failure for kubelet restarts on a new node rather than a fresh one. I switched from crun to runc only to experience the same. 
Currently on 1.21.2. Kubelet constantly restarts with the following alongside crio configured to use crun or runc.
```sh
Jul 15 03:31:45 manjpc kubelet[10668]: E0715 03:31:45.096605   10668 node_container_manager_linux.go:57] "Failed to create cgroup" err="error while starting unit \"kubepods.slice\" with properties [{Name:Description Value:\"libcontainer container kubepods.slice\"} {Name:Wants Value:[\"-.slice\"]} {Name:MemoryAccounting Value:true} {Name:CPUAccounting Value:true} {Name:IOAccounting Value:true} {Name:TasksAccounting Value:true} {Name:DefaultDependencies Value:false}]: Unit kubepods.slice already exists." cgroupName=[kubepods]
Jul 15 03:31:45 manjpc kubelet[10668]: E0715 03:31:45.096631   10668 kubelet.go:1384] "Failed to start ContainerManager" err="error while starting unit \"kubepods.slice\" with properties [{Name:Description Value:\"libcontainer container kubepods.slice\"} {Name:Wants Value:[\"-.slice\"]} {Name:MemoryAccounting Value:true} {Name:CPUAccounting Value:true} {Name:IOAccounting Value:true} {Name:TasksAccounting Value:true} {Name:DefaultDependencies Value:false}]: Unit kubepods.slice already exists."
```
Deleting kubepods.slice via `systemctl stop kubepods.slice` or adding `--cgroups-per-qos=false --enforce-node-allocatable=""` to kubelet args doesn't resolve this issue. 
With an exception...
It previously occurred on an SoC of mine running the same version, and adding those flags did the trick. This was the same solution minikube applied https://github.com/kubernetes/minikube/pull/11258 which related to https://github.com/kubernetes/kubernetes/issues/43856, though that machine is a dev arm board, running a branch of a kernel among other things alongside crio and crun. So I'm still having trouble tracking down what's causing it.

**Comment by odinuge** [comment]
> @odinuge I did not see it as that critical, thought of it more as some nice-to-have but not really required-right-now set of fixes.
> 
> Can you elaborate on what do you think is the exact problem that opencontainers/runc#3082 fixes for kubernetes? Temporarily freezing the sub-cgroups of the pod? If this is it, than it's not a new issue -- it is there since v1.0.0-rc91~48 (opencontainers/runc@afe8348).

Not sure what you mean by that. In k8s `master` it does _not_ freeze. Before last runc bump we didn't use the systemd manager `Set` to update the value, so in that case we had no freeze either.

The big issue is that if k8s keep freezing `kubepods.slice`, `burstable.slice`, `pod-xyz.slice` etc., and users use `runc v1.0.0` (or older, or other runtimes with the same behavior), their containers will permanently freeze. Due to the nature of that, debugging it and finding the cause is a big pain, since everything just "times out" without any meaningful log output. 

> If it is needed, we can definitely backport it to 1.0 and speed up the review. Could you please add a comment to that PR saying you think it needs to be backported to 1.0?

If we can get that into `v1.0.1` and get a release quite quickly, I am optimistic we can get it merged while still having some time for it to soak. That would be nice. I don't have the release schedule of `v1.21` in my head right now tho. (if we can hit next release).

> Another candidate is opencontainers/runc#3067 -- please let us know if you feel it's needed.

Yeah, that one would also definitely be nice to get as soon as possible. It is not that critical as for cgroup v1, since the freezing issue fixed in https://github.com/opencontainers/runc/pull/3081 only applies to cgroup v1. But yeah, getting that into `v1.0.1`, and merged in k8s as soon as possible would be nice. Having to freeze on cgroup v2 feels quite annoying either way.

> If this is not the best place to discuss it, let's move on to opencontainers/runc#3076

That is definitely possible. Not sure what the cherry-picking rules are for runc. I thought only critical regressions and CVEs were cherry-picked, and for `runc` itself https://github.com/opencontainers/runc/pull/3081 isn't any of that. But if you want to cherry-pick the two extra (https://github.com/opencontainers/runc/pull/3081 and opencontainers/runc#3067) that would be nice. Although there wouldn't be that much difference compared to getting a `v1.1.0` release.

**Comment by odinuge** [comment]
The best overall would be to get reviews/approvals to PRs connected to https://github.com/kubernetes/kubernetes/issues/102835 and https://github.com/kubernetes/kubernetes/issues/102836 (and https://github.com/kubernetes/kubernetes/issues/102394) so that we can get better CI signals. That would help tremendously. (and everything related to https://github.com/kubernetes/kubernetes/issues/102148)

**Comment by ehashman** [comment]
@sfxworks thanks for the report, we believe that this currently affects 1.21.2 (but not prior versions of 1.21) and the 1.22 release candidate based on our RCA. It's a release blocker, we're working on it.

**Comment by ehashman** [comment]
@odinuge Got https://github.com/kubernetes/test-infra/pull/22553 approved to help out with the failing serial tests on cri-o/systemd driver, https://github.com/kubernetes/kubernetes/pull/102169 is also pending and should help.

**Comment by kolyshkin** [comment]
@odinuge thanks! We're backporting both https://github.com/opencontainers/runc/pull/3082 and https://github.com/opencontainers/runc/pull/3067 to 1.0.1, and still hope to have a release soon.

**Comment by kolyshkin** [comment]
@odinuge @ehashman @dims PTAL https://github.com/kubernetes/kubernetes/pull/103743

**Comment by ihgann** [comment]
Can confirm I just experienced this as well after upgrading to `v1.21.2`. It's intermittent as well.

**Comment by arren-ru** [comment]
Not sure, but problem still exists in both 1.21.2 and 1.21.3 versions.
Stuck with *.slice problem: not creating container and complaining that slice exists
Also found fatal error which causes failure of kubelet at all:
```
kubelet.go:1384] "Failed to start ContainerManager" err="failed to initialize top level QOS containers: failed to create top level Burstable QOS cgroup : error while starting unit \"kubepods-burstable.slice\" with properties [{Name:Description Value:\"libcontainer container kubepods-burstable.slice\"} {Name:Wants Value:[\"kubepods.slice\"]} {Name:MemoryAccounting Value:true} {Name:CPUAccounting Value:true} {Name:IOAccounting Value:true} {Name:TasksAccounting Value:true} {Name:DefaultDependencies Value:false}]: Unit kubepods-burstable.slice already exists."

kubelet.service: Main process exited, code=exited, status=1/FAILURE
```

**Comment by sfxworks** [comment]
Yeah, can this issue be reopened? For rook 1.6, 
This works
```yaml
    - name: "store-1"
      devices:
      - name: /dev/disk/by-id/scsi-3600508b1001c0f782e68f9977ff4afee
        config:
          diskType: hdd
          metadataDevice: nvme0n1
```
Though this does not
```yaml
    - name: "store-1"
      devices:
      - name: /dev/disk/by-id/scsi-3600508b1001c0f782e68f9977ff4afee
        config:
          diskType: hdd
          metadataDevice: /dev/disk/by-id/nvme-WDBRPG0010BNC-WRSN_21073L802226
```
```
lrwxrwxrwx 1 root root 13 Jul 28 12:35 /dev/disk/by-id/nvme-WDBRPG0010BNC-WRSN_21073L802226 -> ../../nvme0n1
```

**Comment by liggitt** [comment]
still an issue with runc 1.0.1 according to https://github.com/kubernetes/kubernetes/pull/103743#issuecomment-896339048

**Comment by odinuge** [comment]
> still an issue with runc 1.0.1 according to #103743 (comment)

The "new" issue/regression is a bit different. Created https://github.com/kubernetes/kubernetes/issues/104280 to make it a bit more clear.

**Comment by liggitt** [comment]
Thanks for opening #104280.

if I'm understanding correctly, this issue (#102676) describes a problem that is still present in release-1.21 (runc rc95+), but we can't pick up the runc version that fixes it (1.0.1) because of #104280?

**Comment by odinuge** [comment]
> if I'm understanding correctly, this issue (#102676) describes a problem that is still present in release-1.21 (runc rc95+), but we can't pick up the runc version that fixes it (1.0.1) because of #104280?

Yeah, that is right.

Imo. both this and #104280 are quite critical, and makes k8s unusable in a production environment, so it is a bit like choosing between two bad issues. The especially bad thing about #104280 is that user will not see any logs indicating that things are critically wrong, while this issue at least gives some context.

**Comment by sfxworks** [comment]
For some reason, even with the latest versions of runc or crun, I ended up having to upgrade kubelet to 1.22.1 to get things to work. This is alongside my previous fixes, such as removing cpu_manager_state.

**Comment by kolyshkin** [comment]
> For some reason, even with the latest versions of runc or crun

This is not about runc or crun as a binary, this is about runc libraries (libcontainer) imported/used by kubelet.

**Comment by sfxworks** [comment]
Oooh I see. Thanks for clarifying.

**Comment by voigt** [comment]
Hi @odinuge, @liggitt, @dims, @kolyshkin 
Bug-Triage here 👋!

I noticed that this ticket is still labeled for 1.22 and as a release-blocker. As 1.22 is already released I'm moving it to v1.23.

Furthermore, I'd like to reconsider whether this bug is an actual release blocker for 1.23 or whether we can remove that label as well...?

/milestone v1.23

**Comment by odinuge** [comment]
> I noticed that this ticket is still labeled for 1.22 and as a release-blocker. As 1.22 is already released I'm moving it to v1.23.

This was fixed by #104530, so it should be ok to close it. It has had some time to soak now, and things look ok. Sorry for the confusion.

/close
